astroid/inference.py

```diff
@@ -59,7 +59,15 @@
 # .infer method ###############################################################
 
 
-def infer_end(self, context=None):
+_InferEndNodeT = TypeVar(
+    "_InferEndNodeT",
+    bound=Union[nodes.Module, nodes.ClassDef, nodes.Lambda, nodes.Const, nodes.Slice],
```

Comment:
`infer_end` would work with any object, regardless if it's a node or not. A bare TypeVar would probably be better.

```py
_T = TypeVar("_T")
```